File: library/src/main/java/io/github/pelmenstar1/rangecalendar/MoveButtonDrawable.java
package io.github.pelmenstar1.rangecalendar;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.Choreographer;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.jetbrains.annotations.NotNull;

final class MoveButtonDrawable extends Drawable {
    private static final class AnimationState {
        private int startColor;
        private int endColor;

        private boolean isRunning;
        private boolean isCancelledBecauseOfQueue;

        public void set(@NotNull AnimationState other) {
            set(other.startColor, other.endColor);
        }

        public void set(int startColor, int endColor) {
            this.startColor = startColor;
            this.endColor = endColor;
        }
    }

    public static final int DIRECTION_LEFT = 0;
    public static final int DIRECTION_RIGHT = 1;

    public static final int ANIM_TYPE_ARROW_TO_CLOSE = 0;
    public static final int ANIM_TYPE_VOID_TO_ARROW = 1;

    private static final int MIN_VISIBLE_ALPHA = 40;

    private static final int[] ENABLED_STATE = new int[]{android.R.attr.state_enabled};
    private static final Choreographer choreographer = Choreographer.getInstance();

    private final ColorStateList colorList;
    private final Paint paint;

    private int color;

    private final int size;
    private final float strokeWidth;

    private float animationFraction;

    private final float[] linePoints = new float[8];
    private int linePointsLength = 8;

    private final int direction;
    private final int animationType;

    private float invStateChangeDuration;
    private long stateChangeStartTime;

    private AnimationState currentAnimationState;
    private AnimationState queuedAnimationState;

    private final Choreographer.FrameCallback stateChangeAnimTickCb = this::onStateChangeAnimTick;
    private final Choreographer.FrameCallback startStateChangeAnimCb = time -> {
        stateChangeStartTime = time;
        setPaintColor(currentAnimationState.startColor);

        choreographer.postFrameCallback(stateChangeAnimTickCb);
    };

    public MoveButtonDrawable(
            @NotNull Context context,
            @NotNull ColorStateList colorList,
            int direction,
            int animType
    ) {
        this.colorList = colorList;
        this.direction = direction;
        this.animationType = animType;

        Resources res = context.getResources();

        strokeWidth = res.getDimension(R.dimen.rangeCalendar_arrowStrokeWidth);
        size = res.getDimensionPixelSize(R.dimen.rangeCalendar_arrowSize);

        color = colorList.getColorForState(ENABLED_STATE, 0);

        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setColor(color);
        paint.setStrokeWidth(strokeWidth);
    }

    public void setAnimationFraction(float fraction) {
        animationFraction = fraction;

        computeLinePoints();
        invalidateSelf();
    }

    public void setStateChangeDuration(long millis) {
        // convert to nanos and then, invert
        invStateChangeDuration = 1f / (float)(1_000_000 * millis);
    }

    private void onStateChangeAnimTick(long time) {
        // currentAnimationState can't be null, because the animation has started
        if (!currentAnimationState.isRunning && currentAnimationState.isCancelledBecauseOfQueue) {
            currentAnimationState.isRunning = true;
            currentAnimationState.isCancelledBecauseOfQueue = false;

            // if current animation state is cancelled because of another state staying in a queue, that
            // means queuedAnimationState can't be null
            currentAnimationState.set(queuedAnimationState);
            stateChangeStartTime = time;

            setPaintColor(currentAnimationState.startColor);

            choreographer.postFrameCallback(stateChangeAnimTickCb);
        } else {
            float fraction = (float) (time - stateChangeStartTime) * invStateChangeDuration;

            if (fraction >= 1f) {
                onStateChangeAnimTickFraction(1f);

                currentAnimationState.isRunning = false;
            } else {
                onStateChangeAnimTickFraction(fraction);

                choreographer.postFrameCallback(stateChangeAnimTickCb);
            }
        }
    }

    private void startStateChangeAnimation(int startColor, int endColor) {
        if (currentAnimationState != null && currentAnimationState.isRunning) {
            currentAnimationState.isRunning = false;
            currentAnimationState.isCancelledBecauseOfQueue = true;

            if (queuedAnimationState == null) {
                queuedAnimationState = new AnimationState();
            }

            queuedAnimationState.set(startColor, endColor);
        } else {
            if(currentAnimationState == null) {
                currentAnimationState = new AnimationState();
            }

            currentAnimationState.set(startColor, endColor);
            choreographer.postFrameCallback(startStateChangeAnimCb);
        }
    }

    private void onStateChangeAnimTickFraction(float fraction) {
        int c = MathUtils.colorLerp(currentAnimationState.startColor, currentAnimationState.endColor, fraction);

        setPaintColor(c);
    }

    private void setPaintColor(int color) {
        paint.setColor(color);
        invalidateSelf();
    }

    private void computeLinePoints() {
        Rect bounds = getBounds();
        int boundsLeft = bounds.left;
        int boundsTop = bounds.top;
        int boundsWidth = bounds.width();
        int boundsHeight = bounds.height();

        int actualLeft = boundsLeft + (boundsWidth - size) / 2;
        int actualTop = boundsTop + (boundsHeight - size) / 2;
        int actualRight = actualLeft + size;
        int actualBottom = actualTop + size;

        int midX = boundsLeft + boundsWidth / 2;
        int midY = boundsTop + boundsHeight / 2;

        float fraction = animationFraction;
        float halfStrokeWidth = strokeWidth * 0.5f;

        float[] points = linePoints;

        if (animationType == ANIM_TYPE_ARROW_TO_CLOSE) {
            float line1EndY = MathUtils.lerp(midY - halfStrokeWidth + 0.5f, actualBottom, fraction);
            float line2EndY = MathUtils.lerp(midY - halfStrokeWidth, actualTop, fraction);

            float line1EndX;
            float line2EndX;
            float anchorX;

            if (direction == DIRECTION_RIGHT) {
                line1EndX = MathUtils.lerp(midX - halfStrokeWidth + 1, actualRight, fraction);
                line2EndX = MathUtils.lerp(midX + halfStrokeWidth, actualRight, fraction);
                anchorX = actualLeft;
            } else {
                line1EndX = MathUtils.lerp(midX - halfStrokeWidth, actualLeft, fraction);
                line2EndX = MathUtils.lerp(midX - halfStrokeWidth, actualLeft, fraction);
                anchorX = actualRight;
            }

            points[0] = anchorX;
            points[1] = actualTop;
            points[2] = line1EndX;
            points[3] = line1EndY;

            points[4] = anchorX;
            points[5] = actualBottom;
            points[6] = line2EndX;
            points[7] = line2EndY;
        } else {
            float anchorX = direction == DIRECTION_LEFT ? actualRight : actualLeft;

            if (fraction <= 0.5f) {
                float sFr = fraction * 2f;

                points[0] = anchorX;
                points[1] = actualBottom;

                points[2] = MathUtils.lerp(anchorX, midX, sFr);
                points[3] = MathUtils.lerp(actualBottom, midY, sFr);

                linePointsLength = 4;
            } else {
                float sFr = fraction * 2f - 1f;

                points[0] = anchorX;
                points[1] = actualBottom;
                points[2] = midX - halfStrokeWidth;
                points[3] = midY - halfStrokeWidth;

                points[4] = midX + halfStrokeWidth - 2.5f;
                points[5] = midY - halfStrokeWidth - 0.5f;
                points[6] = MathUtils.lerp(points[4], anchorX, sFr);
                points[7] = MathUtils.lerp(points[5], actualTop, sFr);

                linePointsLength = 8;
            }
        }
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        computeLinePoints();
    }

    @Override
    protected boolean onStateChange(int[] state) {
        int oldColor = color;
        int newColor = colorList.getColorForState(state, colorList.getDefaultColor());

        if (oldColor != newColor) {
            color = newColor;

            startStateChangeAnimation(oldColor, newColor);

            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean isStateful() {
        return true;
    }

    @Override
    public void draw(@NonNull Canvas c) {
        float[] points = linePoints;
        if (linePointsLength == 4) {
            c.drawLine(points[0], points[1], points[2], points[3], paint);
        } else {
            c.drawLines(points, 0, linePointsLength, paint);
        }
    }

    @Override
    public void setAlpha(int alpha) {
        setPaintColor(ColorHelper.withAlpha(color, alpha));
    }

    @Nullable
    @Override
    public ColorFilter getColorFilter() {
        return paint.getColorFilter();
    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
        paint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        switch (Color.alpha(color)) {
            case 0:
                return PixelFormat.TRANSPARENT;
            case 255:
                return PixelFormat.OPAQUE;
            default:
                return PixelFormat.TRANSLUCENT;
        }
    }
}

File: library/src/main/java/com/github/pelmenstar1/rangecalendar/MoveButtonDrawable.kt
package com.github.pelmenstar1.rangecalendar

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.*
import android.graphics.drawable.Drawable
import androidx.annotation.RestrictTo
import androidx.core.graphics.alpha
import com.github.pelmenstar1.rangecalendar.utils.lerp
import com.github.pelmenstar1.rangecalendar.utils.withCombinedAlpha

/**
 * A [Drawable] object that draws an arrow. The arrow can also be transitioned to a cross.
 *
 * The class is not expected to be used outside the library.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP_PREFIX)
class MoveButtonDrawable(
    context: Context,
    private val colorList: ColorStateList,
    private val direction: Int,
    private val animationType: Int
) : Drawable() {
    internal val arrowPaint: Paint

    private var arrowColor: Int
    private var arrowColorAlpha = 1f

    private var arrowSize = 0f
    private var arrowStrokeWidth: Float

    internal var arrowUsePath = false
    internal val arrowPath = Path()
    internal val arrowLinePoints = FloatArray(4)

    private var arrowAnimFraction = 0f
    private val colorAnimator = MoveButtonDrawableColorAnimator(
        colorCallback = ::setPaintColor
    )

    init {
        val res = context.resources

        arrowStrokeWidth = res.getDimension(R.dimen.rangeCalendar_arrowStrokeWidth)
        arrowSize = res.getDimension(R.dimen.rangeCalendar_arrowSize)

        arrowColor = colorList.getColorForState(ENABLED_STATE, colorList.defaultColor)

        arrowPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            color = arrowColor
            strokeWidth = arrowStrokeWidth

            strokeCap = Paint.Cap.ROUND
            strokeJoin = Paint.Join.ROUND
        }
    }

    fun setAnimationFraction(fraction: Float) {
        arrowAnimFraction = fraction

        computeLinePoints()
        invalidateSelf()
    }

    fun setStateChangeDuration(millis: Long) {
        colorAnimator.duration = millis
    }

    fun setArrowSize(value: Float) {
        arrowSize = value

        computeLinePoints()
        invalidateSelf()
    }

    fun setArrowStrokeWidth(value: Float) {
        arrowStrokeWidth = value
        arrowPaint.strokeWidth = value

        computeLinePoints()
        invalidateSelf()
    }

    private fun setPaintColor(color: Int) {
        val newColor = color.withCombinedAlpha(arrowColorAlpha)

        if (arrowPaint.color != newColor) {
            arrowPaint.color = newColor

            invalidateSelf()
        }
    }

    private fun computeLinePoints() {
        val bounds = bounds
        if (bounds.isEmpty) {
            return
        }

        val fraction = arrowAnimFraction
        val path = arrowPath
        val linePoints = arrowLinePoints

        val adjOffset = arrowStrokeWidth * 0.5f
        val halfArrowSize = arrowSize * 0.5f
        val adjHalfArrowSize = halfArrowSize - adjOffset

        val midX = (bounds.left + bounds.right) * 0.5f
        val midY = (bounds.top + bounds.bottom) * 0.5f

        val actualLeft = midX - adjHalfArrowSize
        val actualTop = midY - adjHalfArrowSize
        val actualRight = midX + adjHalfArrowSize
        val actualBottom = midY + adjHalfArrowSize

        val anchorX: Float
        val invAnchorX: Float

        if (direction == DIRECTION_LEFT) {
            anchorX = actualRight
            invAnchorX = actualLeft
        } else {
            anchorX = actualLeft
            invAnchorX = actualRight
        }

        // The drawable is animated from the arrow to the cross.
        if (animationType == ANIM_TYPE_ARROW_TO_CROSS) {
            arrowUsePath = true

            // If fraction is 0, it means that we're drawing simple arrow. We should render the arrow as
            // a polygonal chain. Rendering it as two discrete lines is not perfect as there will be an artefact
            // where the lines connect.
            if (fraction == 0f) {
                path.apply {
                    rewind()

                    moveTo(anchorX, actualTop)
                    lineTo(midX, midY)
                    lineTo(anchorX, actualBottom)
                }
            } else {
                // Gradually animate the arrow to the cross.
                val delta = adjHalfArrowSize * fraction

                val line1EndY = midY + delta
                val line2EndY = midY - delta

                val lineEndX = lerp(midX, invAnchorX, fraction)

                path.apply {
                    rewind()

                    moveTo(anchorX, actualTop)
                    lineTo(lineEndX, line1EndY)

                    moveTo(anchorX, actualBottom)
                    lineTo(lineEndX, line2EndY)
                }
            }
        } else {
            // The drawable is animated from void (nothing) to the arrow.
            if (fraction <= 0.5f) {
                // If fraction <= 0.5, it means that we have a line that starts from the bottom (left/right depending on direction)
                // and the end of the line is animated to the center.

                // When fraction = 0.5, the end of the line should be at the center.
                // So we need to scale the fraction from [0; 0.5] to [0; 1]
                val scaledFraction = fraction * 2f

                val lineEndX = lerp(anchorX, midX, scaledFraction)
                val lineEndY = lerp(actualBottom, midY, scaledFraction)

                // We can simply draw a line instead of using Path.
                arrowUsePath = false

                linePoints[0] = anchorX
                linePoints[1] = actualBottom
                linePoints[2] = lineEndX
                linePoints[3] = lineEndY
            } else {
                // Now we have a line that starts from the bottom (left/right depending on the direction) and ends in the center
                // and a line that starts from the center and the end of that line is animated to the top (left/right depending on the direction)

                // Scale the fraction from [0.5; 1] to [0; 1]
                val scaledFraction = fraction * 2f - 1f

                val lineEndX = lerp(midX, anchorX, scaledFraction)
                val lineEndY = lerp(midY, actualTop, scaledFraction)

                arrowUsePath = true

                path.apply {
                    rewind()

                    moveTo(anchorX, actualBottom)
                    lineTo(midX, midY)
                    lineTo(lineEndX, lineEndY)
                }
            }
        }
    }

    override fun onBoundsChange(bounds: Rect) {
        computeLinePoints()
    }

    override fun onStateChange(state: IntArray): Boolean {
        val oldColor = arrowColor
        val newColor = colorList.getColorForState(state, colorList.defaultColor)

        return if (oldColor != newColor) {
            arrowColor = newColor
            colorAnimator.start(oldColor, newColor)

            true
        } else false
    }

    override fun isStateful() = true

    override fun draw(c: Canvas) {
        val paint = arrowPaint

        if (arrowUsePath) {
            c.drawPath(arrowPath, paint)
        } else {
            val points = arrowLinePoints

            c.drawLine(points[0], points[1], points[2], points[3], paint)
        }

        if (RENDER_BOUNDS) {
            // It's only used for debugging purposes. Allocating is OK.
            val infoPaint = Paint().apply {
                style = Paint.Style.STROKE
                color = Color.RED
                strokeWidth = 1f
            }

            val bounds = bounds

            val cx = (bounds.left + bounds.right) * 0.5f
            val cy = (bounds.top + bounds.bottom) * 0.5f

            val halfSize = arrowSize * 0.5f

            val left = cx - halfSize
            val top = cy - halfSize
            val right = cx + halfSize
            val bottom = cy + halfSize

            c.drawRect(left, top, right, bottom, infoPaint)
            c.drawLine(left, top, right, bottom, infoPaint)
            c.drawLine(right, top, left, bottom, infoPaint)
        }
    }

    override fun setAlpha(alpha: Int) {
        setAlpha(alpha / 255f)
    }

    fun setAlpha(alpha: Float) {
        if (arrowColorAlpha != alpha) {
            arrowColorAlpha = alpha

            colorAnimator.alpha = alpha

            // arrowColorAlpha is changed, paint color should be updated too.
            setPaintColor(arrowColor)
        }
    }

    override fun getColorFilter(): ColorFilter? = arrowPaint.colorFilter

    override fun setColorFilter(colorFilter: ColorFilter?) {
        arrowPaint.colorFilter = colorFilter
    }

    @Deprecated("Deprecated in Java")
    override fun getOpacity(): Int {
        return when (arrowColor.alpha) {
            0 -> PixelFormat.TRANSPARENT
            255 -> PixelFormat.OPAQUE
            else -> PixelFormat.TRANSLUCENT
        }
    }

    companion object {
        private const val RENDER_BOUNDS = false

        const val DIRECTION_LEFT = 0
        const val DIRECTION_RIGHT = 1

        const val ANIM_TYPE_ARROW_TO_CROSS = 0
        const val ANIM_TYPE_VOID_TO_ARROW = 1

        private val ENABLED_STATE = intArrayOf(android.R.attr.state_enabled)
    }
}